presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -72,18 +83,33 @@ public synchronized void setResourceOvercommit()
         maxMemory = memoryPool.getMaxBytes();
     }
 
+    public synchronized boolean isWaitingForRevocableMemory()
+    {
+        return !revocableFuture.isDone();
+    }
+
     public synchronized ListenableFuture<?> reserveMemory(long bytes)
     {
         checkArgument(bytes >= 0, "bytes is negative");
 
         if (reserved + bytes > maxMemory) {
             throw exceededLocalLimit(succinctBytes(maxMemory));
         }
-        ListenableFuture<?> future = memoryPool.reserve(queryId, bytes);
+
+        ListenableFuture<?> future = reserveMemoryUnsafe(bytes);
         reserved += bytes;
         return future;
     }
 
+    public synchronized ListenableFuture<?> reserveRevocableMemory(long bytes)
+    {
+        checkArgument(bytes >= 0, "bytes is negative");
+
+        ListenableFuture<?> future = reserveMemoryUnsafe(bytes);
```

Comment:
so it is now possible to allocate above `maxMemory`? Previously exception was being thrown: https://github.com/prestodb/presto/pull/5711/commits/2a4d2197501b59e7c74ff3b863ff88483de7ab4e#diff-68ac3b293142f2c6f9d1f477192b9634L79.

I think we should wait for `revocableFuture` and make sure then we have enough free memory.